### PR TITLE
elimination parsing, syntax & core term

### LINKED ISSUE #24: pattern matching
## Pattern matching
As the opposite of construction, one should be able to destruct the result of data type. For example,
```
(+) : Nat -> Nat
zero + n = n
(suc m') + n = suc (m' + n)
```

Of course, the code would only handle trivial pattern for now, which is Weak-head constructor with following variable-patterns.

And notice the top-level syntax haven't done, and an elimination syntax still needed, so we will have syntax like

```
a + b = elim a
  | zero => b
  | suc n => suc $ n + b
```

## References
1. https://dl.acm.org/doi/pdf/10.1145/3408983
2. http://conal.net/papers/elliott90.pdf

### PR DESCRIPTION
**Pull request by CYBAI**
This PR will introduce the syntax of `elim`ination and also implement the parser for it.

When this PR is merged, users would be able to write code like

```
let zero? : Nat → Bool =
\n => elim n
| zero => True
| suc _ => False
;
```

and do pattern matching based on elimination.

related #24

### CODE REVIEW

**Review by dannypsnl — CHANGES_REQUESTED** [review]
You have to use relate #24 here, and add a description

**Comment by dannypsnl** [comment]
Else the code is ready to merge

File: src/Violet/Core.idr [hunk 1]
```diff
@@ -101,6 +102,7 @@ mutual
           newCtx = extendCtx emptyCtx x a'
       (ty, restEnvAndCtx) <- infer' (newEnv <+> env) (newCtx <+> ctx) u
       pure (ty, (newEnv, newCtx) <+> restEnvAndCtx)
+    Elim t cases => ?todo2
```
Comment (rayshih) [inline src/Violet/Core.idr:105]: I assume this will be implemented later in the same PR right?

Comment (dannypsnl) [reply src/Violet/Core.idr:105]: There should have another PR, else would have too many conflict

Comment (dannypsnl) [reply src/Violet/Core.idr:105]: Let me convert this PR to ready first

Comment (CYBAI) [reply src/Violet/Core.idr:105]: yeah, let's make this PR only for syntax and parser and let's implement `eval` and `infer` later? WDYT?

**Pull request merged**